subprojects/GOPATH/src/github.com/microsoft/hivedscheduler/pkg/algorithm/hived_algorithm.go

```diff
@@ -275,87 +316,273 @@ func (h *HivedAlgorithm) GetAffinityGroup(name string) api.AffinityGroup {
+		c.GetAPIStatus().VirtualCell.CellHealthiness = api.CellBad
+		c.virtualCell.GetAPIStatus().CellHealthiness = api.CellBad
+		c.virtualCell.GetAPIStatus().PhysicalCell.CellHealthiness = api.CellBad
+	}
+	chain, level := c.GetChain(), c.GetLevel()
+	h.numBadCells[chain][level]++
+	if h.totalNumVCFreeCells[chain][level] > h.numLeftCells[chain][level]-h.numBadCells[chain][level] {
+		klog.Warningf("Cell type %v (chain %v level %v) now has fewer healthy cells (healthy %v, bad %v) "+
+			"than the total free cells of all the VCs (%v). Certain VCs may be allocated bad cells.",
+			h.cellTypes[chain][level], chain, level, h.numLeftCells[chain][level]-h.numBadCells[chain][level],
+			h.numBadCells[chain][level], h.totalNumVCFreeCells[chain][level])
+		for vcName, vcQuota := range h.numVCFreeCells {
+			if vcQuota[chain][level] > h.numLeftCells[chain][level]-h.numBadCells[chain][level] {
+				klog.Warningf("Cell type %v (chain %v level %v) now has fewer healthy cells (healthy %v, bad %v) "+
+					"than the free cells of the VC %v (%v). The VC would expect being allocated bad cells.",
+					h.cellTypes[chain][level], chain, level,
+					h.numLeftCells[chain][level]-h.numBadCells[chain][level], h.numBadCells[chain][level],
+					vcName, h.totalNumVCFreeCells[chain][level])
+				for _, vc := range h.vcSchedulers[vcName].getNonReservedFreeCellList()[chain][level] {
+					virtualCell := vc.(*VirtualCell)
+					if virtualCell.GetPhysicalCell() == nil && virtualCell.GetAPIStatus().CellHealthiness != api.CellBad {
+						virtualCell.GetAPIStatus().CellHealthiness = api.CellBad
+						break
+					}
+				}
+			}
+		}
+	}
+
+	if parent := c.GetParent(); parent != nil {
+		for _, buddy := range parent.GetChildren() {
+			if buddy.(*PhysicalCell).GetAPIStatus().CellHealthiness != api.CellBad {
+				return
+			}
+		}
+		h.setBadCell(parent.(*PhysicalCell))
+	}
+}
+
+// setHealthyCell marks a physical cell as healthy and recursively for its parent, guaranteeing that
+// a cell is healthy if any of its children is healthy. The virtual cell that this physical cell
+// is bound to will also be marked as healthy.
+func (h *HivedAlgorithm) setHealthyCell(c *PhysicalCell) {
+	newState := api.CellFree
+	if c.GetPriority() != freePriority {
+		newState = api.CellUsed
+	}
+	c.GetAPIStatus().CellState = newState
+	if c.virtualCell != nil {
+		c.virtualCell.GetAPIStatus().CellState = newState
+		c.virtualCell.GetAPIStatus().PhysicalCell.CellState = newState
+		c.GetAPIStatus().VirtualCell.CellState = newState
+	}
+	chain, level := c.GetChain(), c.GetLevel()
+	h.numBadCells[chain][level]--
+	if h.totalNumVCFreeCells[chain][level] >= h.numLeftCells[chain][level]-h.numBadCells[chain][level] {
+		if h.totalNumVCFreeCells[chain][level] == h.numLeftCells[chain][level]-h.numBadCells[chain][level] {
+			klog.Infof("Cell type %v (chain %v level %v) now has sufficient healthy cells for "+
+				"allocating the total free cells of all the VCs (quota %v, healthy %v, bad %v).",
```

Comment:
>quota [](start = 53, length = 5)

Rename quota to be a more concert concept. Such as total/free cells

Reply:
Also for other places:

C %v (quota %v, healthy %v, bad %

---
In reply to: [389240713](https://github.com/microsoft/pai/pull/4153#discussion_r389240713) [](ancestors = 389240713)